pkg/frontend/middleware/audit.go

```diff
@@ -0,0 +1,132 @@
+func Audit(env env.Core, entry *logrus.Entry) func(http.Handler) http.Handler {
+	return func(h http.Handler) http.Handler {
+		return http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+			callerIdentity, callerType, correlationID, requestID := callerRequestData(r)
+			targetResourceName, targetResourceType := targetResourceData(entry, r)
+
+			entry = entry.WithFields(logrus.Fields{
+				audit.EnvKeyAppID:             audit.SourceRP,
+				audit.EnvKeyCloudRole:         audit.CloudRoleRP,
+				audit.EnvKeyCorrelationID:     correlationID,
+				audit.EnvKeyEnvironment:       env.Environment().Name,
+				audit.EnvKeyHostname:          env.Hostname(),
+				audit.EnvKeyLocation:          env.Location(),
+				audit.PayloadKeyCategory:      audit.CategoryResourceManagement,
+				audit.PayloadKeyOperationName: fmt.Sprintf("%s %s", r.Method, r.URL.Path),
+				audit.PayloadKeyRequestID:     requestID,
+				audit.PayloadKeyCallerIdentities: []audit.CallerIdentity{
+					{
+						CallerIdentityType:  callerType,
+						CallerIdentityValue: callerIdentity,
+						CallerIPAddress:     r.RemoteAddr,
+					},
+				},
+				audit.PayloadKeyTargetResources: []audit.TargetResource{
+					{
+						TargetResourceName: targetResourceName,
+						TargetResourceType: targetResourceType,
+					},
+				},
+			})
+
+			defer func() {
+				resultType, resultDescription := resultData(w)
+				entry.WithFields(logrus.Fields{
+					audit.PayloadKeyResult: audit.Result{
+						ResultType:        resultType,
+						ResultDescription: resultDescription,
+					},
+				}).Info("audit event")
+			}()
+
+			h.ServeHTTP(w, r)
+		})
+	}
+}
+
+func callerRequestData(r *http.Request) (
+	string, string, string, string) {
+
+	var (
+		callerIdentity = r.UserAgent()
+		callerType     = audit.CallerIdentityTypeApplicationID
+		correlationID  = ""
+		requestID      = ""
+	)
+
+	// if log middleware isn't run, this type assertion will panic as intended.
+	// the frontend will recover from the panic and exit
+	correlationData := r.Context().Value(ContextKeyCorrelationData).(*api.CorrelationData)
+	if correlationData != nil {
+		if correlationData.ClientPrincipalName != "" {
+			callerIdentity = correlationData.ClientPrincipalName
+			callerType = audit.CallerIdentityTypeObjectID
+		}
+
+		correlationID = correlationData.CorrelationID
+		requestID = correlationData.RequestID
+	}
+
+	return callerIdentity, callerType, correlationID, requestID
+}
+
+func targetResourceData(entry *logrus.Entry, r *http.Request) (string, string) {
+	matches := utillog.RXTolerantResourceID.FindStringSubmatch(r.URL.Path)
+	if matches == nil {
+		return audit.UnknownValue, audit.UnknownValue
+	}
+
+	var resourceName, resourceKind string
+	if matches[3] != "" {
+		resourceKind = matches[3]
+	}
+
+	if matches[5] != "" {
+		resourceName = fmt.Sprintf("/subscriptions/%s/resourcegroups/%s/providers/%s/%s/%s", matches[1], matches[2], matches[3], matches[4], matches[5])
```

Comment:
Once the decision on what to do here is finalised, it needs to be encoded into a unit test.

Reply:
ah... the confusion here is that i intended this to be the resource name (i.e. cluster name), not the resource ID. The resource ID is captured in the operation name in the `<http_verb> <resource_id>` format. 

in the scenarios you described, the resource name will be assigned an empty string since there isn't one.  e.g.,
```
az aro show -g isim-dev -n isim-dev-cluster
...
{
  "OperationName": "GET /subscriptions/subid/resourcegroups/aro-isim/providers/microsoft.redhatopenshift/openshiftclusters/aro-isim-cluster",
  "TargetResources":[{"TargetResourceType":"microsoft.redhatopenshift","TargetResourceName":"aro-isim-cluster"}]
}

az aro list -g isim-dev
...
{
   "OperationName":"GET /subscriptions/subid/resourcegroups/aro-isim/providers/microsoft.redhatopenshift/openshiftclusters",
   "TargetResources":[{"TargetResourceType":"microsoft.redhatopenshift","TargetResourceName":\"\"}]}
}
```

the value can be anything we want it to be. unless you have other thoughts on this, my lean towards just:

```
if matches[5] != "" {
		resourceName = matches[5]
}
```